cartridge/failover.lua

```diff
@@ -200,10 +198,150 @@ local function apply_config(mod)
     )
 end
 
-local function reconfigure_all()
+local function constitute_oneself(active_leaders, opts)
+    checks('table', {
+        timeout = 'number',
+    })
+
+    local confapplier = require('cartridge.confapplier')
+    local instance_uuid = confapplier.get_instance_uuid()
+    local replicaset_uuid = confapplier.get_replicaset_uuid()
+
+    local topology_cfg = vars.clusterwide_config:get_readonly('topology')
+
+    if active_leaders[replicaset_uuid] ~= instance_uuid then
+        -- I'm not a leader
+        vars.cache.is_vclockkeeper = false
+        vars.cache.is_leader = false
+        vars.cache.is_rw = topology_cfg.replicasets[replicaset_uuid].all_rw
+        return true
+    end
+
+    -- Get ready to become a leader
+    if not vars.consistency_needed then
+        vars.cache.is_vclockkeeper = false
+        vars.cache.is_leader = true
+        vars.cache.is_rw = true
+        return true
+    elseif vars.cache.is_vclockkeeper then
+        -- I'm already a vclockkeeper
+        return true
+    end
+
+    local deadline = fiber.time() + opts.timeout
+
+    -- Go to the external state provider
+    local session = assert(vars.client):get_session()
+    if session == nil or not session:is_alive() then
+        return nil, StateProviderError:new('State provider unavailable')
+    end
+
+    -- Query current vclockkeeper
+    -- WARNING: implicit yield
+    local vclockkeeper, err = session:get_vclockkeeper(replicaset_uuid)
+    if err ~= nil then
+        return nil, ConsistentSwitchoverError:new(err)
+    end
+    fiber.testcancel()
+
+    if vclockkeeper == nil then
+        -- It's absent, no need to wait anyone
+        goto set_vclockkeeper
+    elseif vclockkeeper.instance_uuid == instance_uuid then
+        -- It's me already
+        vars.cache.is_vclockkeeper = true
+        vars.cache.is_leader = true
+        vars.cache.is_rw = true
+        return true
+    end
+
+    do
+        -- Go to the vclockkeeper and query its LSN
+        local vclockkeeper_uuid = vclockkeeper.instance_uuid
+        local vclockkeeper_srv = topology_cfg.servers[vclockkeeper_uuid]
+        if vclockkeeper_srv == nil then
+            return nil, ConsistentSwitchoverError:new('Alien vclockkeeper')
+        end
+
+        local vclockkeeper_uri = vclockkeeper_srv.uri
+
+        local timeout = deadline - fiber.time()
+        -- WARNING: implicit yield
+        local vclockkeeper_info, err = errors.netbox_eval(
+            pool.connect(vclockkeeper_uri, {wait_connected = false}), [[
```

Comment:
It's a bit tricky. Why don't you use predefined stored procedure instead of raw eval?

Reply:
BTW
```lua
tarantool> x = [[
         >             box.ctl.wait_ro(...)
         >             return {
         >                 id  = box.info.id,
         >                 lsn = box.info.lsn,
         >             }
         >         ]]
---
...

tarantool> #x
---
- 147
...
```

If you want to to preserve it as is you can predefine it and save it with smaller length

Reply:
And you don't check that `wait_ro ` returns you.

Reply:
Moved it to a special stored procedure. But there's nothing to check - `wait_ro` doesn't return anything.